### PR TITLE
Filter Hudi partitions from metastore based on constraint predicate

### PR DESCRIPTION
**Pull request by codope**
Fixes https://github.com/apache/hudi/issues/8485

Test plan - (Please fill in how you tested your changes)

- Added a unit test.
- Tested a query with `>=` partition predicate and compared plans (check number of rows in `Fragment 2` in the below plans).

Query: `select count(1) from inventory where inv_date_sk>=2451676;`

Before this fix:
```
 Fragment 1 [SINGLE]
     CPU: 103.25ms, Scheduled: 194.86ms, Input: 261 rows (2.29kB); per task: avg.: 261.00 std.dev.: 0.00, Output: 1 row (9B)
     Output layout: [count]
     Output partitioning: SINGLE []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     - Aggregate(FINAL) => [count:bigint]
             CPU: 9.00ms (0.46%), Scheduled: 11.00ms (0.03%), Output: 1 row (9B)
             Input avg.: 261.00 rows, Input std.dev.: 0.00%
             count := "presto.default.count"((count_4)) (1:24)
         - LocalExchange[SINGLE] () => [count_4:bigint]
                 CPU: 17.00ms (0.86%), Scheduled: 19.00ms (0.05%), Output: 261 rows (2.29kB)
                 Input avg.: 16.31 rows, Input std.dev.: 245.74%
             - RemoteSource[2] => [count_4:bigint]
                     CPU: 18.00ms (0.91%), Scheduled: 60.00ms (0.15%), Output: 261 rows (2.29kB)
                     Input avg.: 16.31 rows, Input std.dev.: 245.74%

 Fragment 2 [SOURCE]
     CPU: 1.96s, Scheduled: 31.81s, Input: 11745000 rows (4.08MB); per task: avg.: 11745000.00 std.dev.: 0.00, Output: 261 rows (2.29kB)
     Output layout: [count_4]
     Output partitioning: SINGLE []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     - Aggregate(PARTIAL) => [count_4:bigint]
             CPU: 197.00ms (9.99%), Scheduled: 220.00ms (0.55%), Output: 261 rows (2.29kB)
             Input avg.: 23793.10 rows, Input std.dev.: 94.41%
             count_4 := "presto.default.count"(*) (1:24)
         - ScanFilterProject[table = TableHandle {connectorId='hudi', connectorHandle='default.inventory', layout='Optional[default.inventory]'}, grouped = false, filterPredicate = (inv_date_sk) >= (INTEGER'2451676'), projectLocality = LOCAL] => []
                 CPU: 1.73s (87.77%), Scheduled: 39.99s (99.23%), Output: 6210000 rows (0B)
                 Input avg.: 45000.00 rows, Input std.dev.: 0.00%
                 inv_date_sk := -13:inv_date_sk:int:PARTITION_KEY (1:38)
                 Input: 11745000 rows (4.08MB), Filtered: 47.13%
```
After the fix:
```
 Fragment 1 [SINGLE]
     CPU: 70.04ms, Scheduled: 914.23ms, Input: 138 rows (1.21kB); per task: avg.: 138.00 std.dev.: 0.00, Output: 1 row (9B)
     Output layout: [count]
     Output partitioning: SINGLE []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     - Aggregate(FINAL) => [count:bigint]
             CPU: 7.00ms (0.26%), Scheduled: 104.00ms (0.42%), Output: 1 row (9B)
             Input avg.: 138.00 rows, Input std.dev.: 0.00%
             count := "presto.default.count"((count_4)) (1:24)
         - LocalExchange[SINGLE] () => [count_4:bigint]
                 CPU: 12.00ms (0.45%), Scheduled: 58.00ms (0.23%), Output: 138 rows (1.21kB)
                 Input avg.: 8.63 rows, Input std.dev.: 174.29%
             - RemoteSource[2] => [count_4:bigint]
                     CPU: 15.00ms (0.56%), Scheduled: 196.00ms (0.78%), Output: 138 rows (1.21kB)
                     Input avg.: 8.63 rows, Input std.dev.: 174.29%

 Fragment 2 [SOURCE]
     CPU: 2.68s, Scheduled: 21.44s, Input: 6210000 rows (2.16MB); per task: avg.: 6210000.00 std.dev.: 0.00, Output: 138 rows (1.21kB)
     Output layout: [count_4]
     Output partitioning: SINGLE []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     - Aggregate(PARTIAL) => [count_4:bigint]
             CPU: 197.00ms (9.99%), Scheduled: 220.00ms (0.55%), Output: 261 rows (2.29kB)
             Input avg.: 23793.10 rows, Input std.dev.: 94.41%
             count_4 := "presto.default.count"(*) (1:24)
         - TableScan[TableHandle {connectorId='hudi', connectorHandle='default.inventory', layout='Optional[default.inventory]'}, layout='Optional[default.inventory{domains={inv_date_sk=[ [["2451676", <max>)] ]}}]
                 CPU: 1.32s (49.44%), Scheduled: 21.21s (84.75%), Output: 6210000 rows (0B)
                 Input avg.: 45000.00 rows, Input std.dev.: 0.00%
                 LAYOUT: default.inventory{domains={inv_date_sk=[ [["2451676", <max>)] ]}}
                 inv_date_sk:int:-13:PARTITION_KEY:: [["2451676", "2452635"]]
                 Input: 6210000 rows (2.16MB), Filtered: 0.00%
```
== RELEASE NOTES ==

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java
Comment (codope) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:620]: Could not add `DateTimeZone` at the call site because `joda-time` was replaced by `java.time` in https://github.com/prestodb/presto/commit/bc82daf13276c44547084eae83671e7eb285f4ee 
So, i have added this util method.

**Comment by codope** [comment]
@vinothchandar @7c00  Can you please review this PR?

**Review by vinothchandar — CHANGES_REQUESTED** [review]
Left first set of comments.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java
Comment (7c00) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:619]: nit: as an convention, we tend to add argument checking code in public method, and omit it in private method.

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java
Comment (7c00) [inline presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java:103]: partitionId -> partitionName, to make the names consistent

**Review by 7c00 — CHANGES_REQUESTED** [review]

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java
Comment (codope) [reply presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java:87]: Yes, just once per query called by the split manager.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java
Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:620]: We could do as you suggested but it will entail copying significant code from HiveUtil, which also contains some usage of deprecated api. I think it's better for maintainability to do it the current way.

Added a test in `TestHiveUtil`.

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java
Comment (codope) [reply presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java:117]: yea, so [Domain](https://github.com/prestodb/presto/blob/e476f311dfc6eb5ee9a0aca15e3e0fc36a51fde9/presto-common/src/main/java/com/facebook/presto/common/predicate/Domain.java) contains [ValueSet](https://github.com/prestodb/presto/blob/e476f311dfc6eb5ee9a0aca15e3e0fc36a51fde9/presto-common/src/main/java/com/facebook/presto/common/predicate/ValueSet.java) and this value set can have two implementations `EquatableValueSet` - for a predicate containing a set of values that are uniquely identifiable, and `SortedRangeSet` - for a predicate containing a set of zero or more ranges of the same type. Previously, we missed `SorteedRangeSet` as `metastore.getPartitionNamesByFilter` only checked for equality. Hence, this additional filtering is needed. It is there in Hive connector but we missed in Hudi.

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java
Comment (codope) [reply presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java:108]: This is just a sanity check. We should never land in this block because in `HudiMetadata.getTableLayouts` we always get the constrain with data and partition columns.

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java
Comment (codope) [reply presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java]: Added `TestHudiPartitionManager`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java
Comment (vinothchandar) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:620]: >but it will entail copying significant code from HiveUtil

How so? All the methods you call here like `parsePartitionValue`  are public static right?

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java
Comment (vinothchandar) [reply presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java:87]: Can you confirm any new RPCs incurred?

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java
Comment (vinothchandar) [reply presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java:102]: Is this UT-ed now?

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java
Comment (vinothchandar) [reply presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java:117]: Thanks for the explanation

File: presto-hudi/src/test/java/com/facebook/presto/hudi/TestHudiIntegration.java
Comment (vinothchandar) [reply presto-hudi/src/test/java/com/facebook/presto/hudi/TestHudiIntegration.java:37]: Can we file the issue and link here in the comment. I can help apply hudi label

**Review by vinothchandar — COMMENTED** [review]
Clarifying comments. Can you please ensure these tests are not flaky (run some 30 times in IDE?). Just dotting all i's

**Review by vinothchandar — COMMENTED** [review]
Clarifying comments. Can you please ensure these tests are not flaky (run some 30 times in IDE?). Just dotting all i's

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java
Comment (codope) [reply presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java:87]: No new RPCs. Any subsequent filtering is local.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java
Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:620]: Yeah all relevant methods in this call are public static but they work with `org.joda.time.DateTimeZone`. In order to avoid adding that library in presto-hudi module, i'll have to copy and rewrite those methods to make them work with `java.time.ZoneId`.

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java
Comment (codope) [reply presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java:102]: Yes, the test that has been added in `TestHudiPartitionManager` goes inside this method and subsequent calls to HiveUtil's methods. It tests both branches of the code at line 117.

File: presto-hudi/src/test/java/com/facebook/presto/hudi/TestHudiIntegration.java
Comment (codope) [reply presto-hudi/src/test/java/com/facebook/presto/hudi/TestHudiIntegration.java:37]: Thanks. I've created an issue - https://github.com/prestodb/presto/issues/19651

**Comment by codope** [comment]
> Clarifying comments. Can you please ensure these tests are not flaky (run some 30 times in IDE?). Just dotting all i's

@vinothchandar Responded to your comments. Also, ran all the tests in presto-hudi module 50 times to ensure no flakiness.

File: presto-hudi/src/test/java/com/facebook/presto/hudi/TestHudiIntegration.java
Comment (vinothchandar) [reply presto-hudi/src/test/java/com/facebook/presto/hudi/TestHudiIntegration.java:37]: labelled with hudi

**Review by vinothchandar — APPROVED** [review]
LGTM

**Comment by vinothchandar** [comment]
@7c00 Can you please take a final pass, before we land this? Will timeout in a day or two and land.

**Review by 7c00 — APPROVED** [review]
LGTM

**Review by 7c00 — APPROVED** [review]
LGTM

**Comment by 7c00** [comment]
This patch include changes to the code outside `presto-hudi`, so another approval from @presto/committers is required. @codope

**Comment by codope** [comment]
Ok, thanks for reviewing @7c00. Some of the GH action failures seem unrelated to this patch. Should I just push an empty commit or rebase with master?
@tdcmeehan Could you please review as well? There is one minor changes in `HiveUtil` and everything else is within `presto-hudi` module.

**Review by tdcmeehan — APPROVED** [review]
LGTM

**Pull request merged**
